File: analysis/import/SCMLogParser/src/test/java/de/maibornwolff/codecharta/importer/scmlogparser/parser/svn/SVNLogParserStrategyTest.java
package de.maibornwolff.codecharta.importer.scmlogparser.parser.svn;

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification;
import de.maibornwolff.codecharta.importer.scmlogparser.parser.LogParserStrategy;
import de.maibornwolff.codecharta.importer.scmlogparser.parser.ParserStrategyContractTest;
import org.assertj.core.util.Lists;
import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;

public class SVNLogParserStrategyTest extends ParserStrategyContractTest {

    private static final List<String> FULL_COMMIT = asList(
            "------------------------------------------------------------------------",
            "r2 | TheAuthor | 2017-05-09 19:57:57 +0200 (Tue, 9 May 2017) | 1 line",
            "Changed paths:",
            "   A /trunk/src/Added.java",
            "   M /trunk/src/Modified.java",
            "   D /trunk/src/Deleted.java",
            "the commit message");

    private SVNLogParserStrategy parserStrategy;

    @Before
    public void setup() {
        parserStrategy = new SVNLogParserStrategy();
    }

    @Override
    protected List<String> getFullCommit() {
        return FULL_COMMIT;
    }


    @Override
    protected Stream<String> getTwoCommitsAsStraem() {
        List<String> twoCommits = Lists.newArrayList("------------------------------------------------------------------------");
        twoCommits.addAll(FULL_COMMIT);
        twoCommits.add("------------------------------------------------------------------------");
        twoCommits.addAll(FULL_COMMIT);
        twoCommits.add("------------------------------------------------------------------------");
        return twoCommits.stream();
    }

    @Override
    protected LogParserStrategy getLogParserStrategy() {
        return parserStrategy;
    }

    @Test
    public void parsesFilenameFromFileMetadata() {
        Modification modification = parserStrategy.parseModification("   M /src/srcFolderTest.txt");
        assertThat(modification.getFilename()).isEqualTo("/src/srcFolderTest.txt");
        assertThat(modification.getType()).isEqualTo(Modification.Type.MODIFY);
    }

    @Test
    public void parsesFilenameFromAddedFile() {
        Modification modification = parserStrategy.parseModification("   A /src/srcFolderTest.txt");
        assertThat(modification.getFilename()).isEqualTo("/src/srcFolderTest.txt");
        assertThat(modification.getType()).isEqualTo(Modification.Type.ADD);
    }

    @Test
    public void parsesFilenameFromDeletedFile() {
        Modification modification = parserStrategy.parseModification("   D  /src/srcFolderTest.txt");
        assertThat(modification.getFilename()).isEqualTo("/src/srcFolderTest.txt");
        assertThat(modification.getType()).isEqualTo(Modification.Type.DELETE);
    }

    @Test
    public void parsesFilenameFromReplacedFile() {
        Modification modification = parserStrategy.parseModification("   R  /src/srcFolderTest.txt");
        assertThat(modification.getFilename()).isEqualTo("/src/srcFolderTest.txt");
        assertThat(modification.getType()).isEqualTo(Modification.Type.UNKNOWN);
    }

    @Test
    public void doesNotParseFilenameWithoutADot() {
        Modification modification = parserStrategy.parseModification("   A /innerFolder");
        assertThat(modification.getFilename()).isEmpty();
    }

    @Test
    public void removesStandardSVNFoldersInFilename() {
        Modification modification = parserStrategy.parseModification("   M /trunk/src/srcFolderTest.txt");
        assertThat(modification.getFilename()).isEqualTo("src/srcFolderTest.txt");
    }

    @Test
    public void acceptsSVNLogWithoutEndingDashes() {
        Stream<String> logLinesWithoutEndingDashes = Stream.of("-----------", "commit data");
        Stream<List<String>> commits = logLinesWithoutEndingDashes.collect(parserStrategy.createLogLineCollector());
        assertThat(commits).hasSize(1);
    }

    @Test
    public void ignoresTooShortLines() {
        List<String> lines = Arrays.asList(" A", "B");
        List<Modification> modifications = parserStrategy.parseModifications(lines);
        assertThat(modifications).isEmpty();
    }
}

File: analysis/import/SCMLogParserV2/src/test/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/parser/svn/SVNLogParserStrategyTest.kt
package de.maibornwolff.codecharta.importer.scmlogparser.parser.svn

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification
import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.MetricsFactory
import de.maibornwolff.codecharta.importer.scmlogparser.parser.LogLineParser
import de.maibornwolff.codecharta.importer.scmlogparser.parser.LogParserStrategy
import de.maibornwolff.codecharta.importer.scmlogparser.parser.ParserStrategyContractTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.util.Arrays
import java.util.stream.Stream

class SVNLogParserStrategyTest : ParserStrategyContractTest() {

    private var parserStrategy: SVNLogParserStrategy = SVNLogParserStrategy()

    override val fullCommit: List<String>
        get() = FULL_COMMIT

    override val twoCommitsAsStraem: Stream<String>
        get() {
            val twoCommits = mutableListOf("------------------------------------------------------------------------")
            twoCommits.addAll(FULL_COMMIT)
            twoCommits.add("------------------------------------------------------------------------")
            twoCommits.addAll(FULL_COMMIT)
            twoCommits.add("------------------------------------------------------------------------")
            return twoCommits.stream()
        }

    override val logParserStrategy: LogParserStrategy
        get() = parserStrategy

    @Test
    fun parsesFilenameFromFileMetadata() {
        val modification = parserStrategy.parseModification("   M /src/srcFolderTest.txt")
        assertThat(modification.currentFilename).isEqualTo("src/srcFolderTest.txt")
        assertThat(modification.type).isEqualTo(Modification.Type.MODIFY)
    }

    @Test
    fun parsesFilenameFromAddedFile() {
        val modification = parserStrategy.parseModification("   A /src/srcFolderTest.txt")
        assertThat(modification.currentFilename).isEqualTo("src/srcFolderTest.txt")
        assertThat(modification.type).isEqualTo(Modification.Type.ADD)
    }

    @Test
    fun parsesFilenameFromDeletedFile() {
        val modification = parserStrategy.parseModification("   D  /src/srcFolderTest.txt")
        assertThat(modification.currentFilename).isEqualTo("src/srcFolderTest.txt")
        assertThat(modification.type).isEqualTo(Modification.Type.DELETE)
    }

    @Test
    fun parsesFilenameFromReplacedFile() {
        val modification = parserStrategy.parseModification("   R  /src/srcFolderTest.txt")
        assertThat(modification.currentFilename).isEqualTo("src/srcFolderTest.txt")
        assertThat(modification.type).isEqualTo(Modification.Type.UNKNOWN)
    }

    @Test
    fun doesNotParseFilenameWithoutADot() {
        val modification = parserStrategy.parseModification("   A /innerFolder")
        assertThat(modification.currentFilename).isEmpty()
    }

    @Test
    fun removesStandardSVNFoldersInFilename() {
        val modification = parserStrategy.parseModification("   M /trunk/src/srcFolderTest.txt")
        assertThat(modification.currentFilename).isEqualTo("src/srcFolderTest.txt")
    }

    @Test
    fun acceptsSVNLogWithoutEndingDashes() {
        val logLinesWithoutEndingDashes = Stream.of("------------------------------------------------------------------------", "commit data")
        val commits = logLinesWithoutEndingDashes.collect(parserStrategy.createLogLineCollector())
        assertThat(commits).hasSize(1)
    }

    @Test
    fun ignoresTooShortLines() {
        val lines = Arrays.asList(" A", "B")
        val modifications = parserStrategy.parseModifications(lines)
        assertThat(modifications).isEmpty()
    }

    @Test
    fun parsesSpecialCommit() {
        val parser = LogLineParser(logParserStrategy, MetricsFactory())
        val commitString = mutableListOf(
            "------------------------------------------------------------------------",
            "r156657 | dpagam05 | 2017-01-02 03:12:18 +0100 (Mo, 02 Jan 2017) | 1 line",
            "Changed paths:",
            "    M /src/Modified.java",
            "Task | Increased automaticly build number | builduser01",
            "------------------------------------------------------------------------"
        )
        val commit = parser.parseCommit(commitString)
        assertThat(commit.filenames)
            .containsExactlyInAnyOrder("src/Modified.java")
    }

    companion object {

        private val FULL_COMMIT = mutableListOf(
            "------------------------------------------------------------------------",
            "r2 | TheAuthor | 2017-05-09 19:57:57 +0200 (Tue, 9 May 2017) | 1 line",
            "Changed paths:",
            "   A /trunk/src/Added.java",
            "   M /trunk/src/Modified.java",
            "   D /trunk/src/Deleted.java",
            "the commit message"
        )
    }
}
